### PR TITLE
Blazor based in-browser WebUI

### PR DESCRIPTION
**Pull request by Kuinox**
This is the first iteration of a WebUI using Blazor.  
Got the idea this monday afternoon and got surprised that I managed to do so much stuff in the evening.  

Because I know everyone here doesn't know web techs, I'll do a quick summary of the techs used:  

### WASM

Since a few years, browsers allow to run [WebAssembly](https://en.wikipedia.org/wiki/WebAssembly) which is yet another Intermediate Language for a VM like .NET IL:  
> [..] WebAssembly was initially designed to enable near-native code execution speed in the web browser.  

Mono have been compiled and updated to support WASM, and Browser execution.

### Razor/Blazor

Now, Razor exists in the .NET world since a long time, and allow to render html pages, in a PHP fashion, and was mostly used to do server side rendering until Blazor came.  

Blazor is a new web framework, that basically put Razor, in the browser, and tada, you have a new front framework.  

Blazor has several disadvantage, and advantage.  

Disadvantages are mainly performances issues:  
We need to bring the .NET Runtime in the browser, even with treeshaking/trimming, it make a big bloated page, and slow startup time (severals seconds for the first content to display).  

Except in our case, we want to run a .NET toolchain in the Browser, so we need the .NET Runtime anyway. This negate the biggest disadvantage of blazor.  

Another disadvantage, and this is my personal taste on this one, is that it's filled with the same overengineering you can see in aspnet.  

Now the advantage 2:  
You are partially shielded from the JS ecosystem.
  You still need to interact with it:  
    - It's still html, css, and you run in a browser.
    - Sometimes you need to interact yourself with a JS library if there are no existing community made binding, or if they are too crappy.

You can use a lot of the .NET librairies, lot of them work out of the box or with light configuration, and you write your app with good ol' C#.

### The WebUI architecture

It's not my first time writting a WebIDE, I already made an attempt in the past for a DSL: https://kuinox.github.io/TQLBlazorDemo/

I stripped a lot of the initial Blazor scaffolding, because in our case we have a single page, so no routing, and we don't need a CSS framework.  

The editors are powered by [monaco](https://github.com/microsoft/monaco-editor).  
Monaco is the core of vscode that can fully run on the web (but have some differences with vscode). To not go through the pain of implementing the binding myself, I use [a community made binding](https://github.com/serdarciplak/BlazorMonaco).  

The only changes I made to the Draco compiler was in ScriptingEngine, by splitting the scripting steps and expose methods for transpilation and C# compilation to an assembly.

The App.razor nicely glue all of that together.

### CODE REVIEW

**Comment by Kuinox** [comment]
I know nobody gonna read the PR text 😢

File: src/Draco.Editor.Web/App.razor [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+        _selectedOutputType = e.Value.ToString();
+        _ = UpdateOutput(); // awaiting will block UI thread
+    }
+
+    public void CodeChange()
+    {
+        _ = UpdateOutput(); // awaiting will block UI thread.
+    }
+
+    string _codeCache = "";
```
Comment (LPeter1997) [inline src/Draco.Editor.Web/App.razor:111]: Random field, please specify publicity and factor to the top, maybe below the props

File: src/Draco.Editor.Web/App.razor [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+            inBuffer.Position = 0;
+            byte[] buffer = inBuffer.ToArray();
+            var hash = Convert.ToBase64String(buffer).TrimEnd('=').Replace('+', '-').Replace('/', '_');
+            await JS.InvokeVoidAsync("setHash", new object[] { hash });
+        }
+        var oldOut = Console.Out;
+
+        try
+        {
+
```
Comment (LPeter1997) [inline src/Draco.Editor.Web/App.razor:132]: Stray newline.

File: src/Draco.Editor.Web/App.razor [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+                    throw new InvalidOperationException();
+            }
+
+        }
+        catch (Exception e)
+        {
+            await _outputViewer!.SetValue(e.ToString());
+        }
+
+        async Task IL()
```
Comment (LPeter1997) [inline src/Draco.Editor.Web/App.razor:154]: Maybe a more descriptive method name would be better?

**Review by LPeter1997 — COMMENTED** [review]
Truly amazing work, thank you! I believe this is a super attractive feature to have, especially this early.

**Comment by LPeter1997** [comment]
> I know nobody gonna read the PR text 😢

I did and it was very informative for my very backend-oriented brain

File: src/Draco.Editor.Web/App.razor [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+    MonacoEditor? _dracoEditor { get; set; }
+    MonacoEditor? _outputViewer { get; set; }
+    string? _selectedOutputType { get; set; } = "Run";
+    StandaloneEditorConstructionOptions _dracoEditorOptions = new StandaloneEditorConstructionOptions
+        {
+            AutomaticLayout = true,
+            Language = "rust",
+            Value = @"func main() {
+    println(""Hello!"");
+}"
```
[the remark this answers is not in the document]
Comment (Kuinox) [reply src/Draco.Editor.Web/App.razor:50]: Doesn't work in razor sadly.

[the remark this answers is not in the document]
Comment (LPeter1997) [reply src/Draco.Editor.Web/App.razor:50]: Alright then, didn't know.

File: src/Draco.Editor.Web/App.razor [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+<div class="editors-container">
+    <MonacoEditor Id="draco-editor" @ref="_dracoEditor" ConstructionOptions="DracoMonacoOptions" OnDidChangeModelContent="CodeChange" OnDidInit="OnInit" />
+    <MonacoEditor Id="output-viewer" @ref="_outputViewer" ConstructionOptions="OutputMonacoOptions" />
+</div>
+
+@code {
+    MonacoEditor? _dracoEditor { get; set; }
+    MonacoEditor? _outputViewer { get; set; }
+    string? _selectedOutputType { get; set; } = "Run";
+    StandaloneEditorConstructionOptions _dracoEditorOptions = new StandaloneEditorConstructionOptions
```
[the remark this answers is not in the document]
Comment (Kuinox) [reply src/Draco.Editor.Web/App.razor:44]: > I won't pretend I know what these belong to, maybe a class?

These are linked to the corresponding HTML elements. These one must be in this razor file.

File: src/Draco.Editor.Web/App.razor [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+        => _outputEditorOptions;
+
+    protected override async Task OnInitializedAsync()
+    {
+        await base.OnInitializedAsync();
+        bool isDarkMode = await JS.InvokeAsync<bool>("isDarkMode");
+        await MonacoEditor.SetTheme(isDarkMode ? "vs-dark" : "vs");
+    }
+
+    async Task OnInit(MonacoEditorBase editor)
```
[the remark this answers is not in the document]
Comment (Kuinox) [reply src/Draco.Editor.Web/App.razor:74]: wdym by that ?

File: src/Draco.Editor.Web/App.razor [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+    {
+        string uriString = NavigationManager.Uri;
+        Uri uri = new Uri(uriString);
+        if (!string.IsNullOrWhiteSpace(uri.Fragment))
+        {
+            string strBuffer = new string(uri.Fragment.Skip(1).ToArray());
+            strBuffer = Uri.UnescapeDataString(strBuffer)
+            .Replace('_', '/')
+            .Replace('-', '+')
+            .PadRight(4 * ((strBuffer.Length + 3) / 4), '=');
```
[the remark this answers is not in the document]
Comment (Kuinox) [reply src/Draco.Editor.Web/App.razor:84]: I'll comment it, it's a simple Base64 to Base64 Url Safe conversion.

[the remark this answers is not in the document]
Comment (LPeter1997) [reply src/Draco.Editor.Web/App.razor:84]: Could you also factor it out into a private static function to not to have such logic inlined?

[the remark this answers is not in the document]
Comment (Kuinox) [reply src/Draco.Editor.Web/App.razor:84]: Sadly .NET doesn't support yet Base64UrlSafe, it's planned for .NET 8.
https://github.com/dotnet/runtime/issues/1658

File: src/Draco.Editor.Web/App.razor [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+            var buffer = Convert.FromBase64String(strBuffer);
+            using (MemoryStream inBuffer = new MemoryStream(buffer))
+            using (DeflateStream gzipStream = new DeflateStream(inBuffer, CompressionMode.Decompress))
+            using (StreamReader outReader = new StreamReader(gzipStream, Encoding.UTF8))
```
[the remark this answers is not in the document]
Comment (Kuinox) [reply src/Draco.Editor.Web/App.razor:88]: Conventions at work, we hate the inline using and rarely use it.
Switched to inline.

[the remark this answers is not in the document]
Comment (LPeter1997) [reply src/Draco.Editor.Web/App.razor:88]: There's no reason to hate it, if there are no consequences of using it. Here we are using it for in-memory streams, which is totally fine. For OS-held resources like a file-handle, I'd be more concerned.

File: src/Draco.Editor.Web/App.razor (hunk 1 above)
Comment (Kuinox) [reply src/Draco.Editor.Web/App.razor:111]: Copy paste leftover.

File: src/Draco.Editor.Web/Program.cs [hunk 9]
```diff
@@ -0,0 +1,13 @@
+using Microsoft.AspNetCore.Components.Web;
+using Microsoft.AspNetCore.Components.WebAssembly.Hosting;
+
+namespace Draco.Editor.Web;
+public class Program
```
Comment (LPeter1997) [inline src/Draco.Editor.Web/Program.cs:5]: Please put a newline between the class and the namespace decl.

File: src/Draco.Editor.Web/App.razor [hunk 10]
```diff
@@ -0,0 +1,38 @@
+@using System.IO.Compression;
```
Comment (LPeter1997) [inline src/Draco.Editor.Web/App.razor:1]: Love this minimal Razor file.

File: src/Draco.Editor.Web/App.cs [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+
+        async Task ShowCSharp()
+        {
+            defaultOutputEditorOptions.Language = "csharp";
+            await this.OutputViewer!.UpdateOptions(defaultOutputEditorOptions);
+            var cSharpCode = ScriptingEngine.CompileToCSharpCode(code);
+            await this.OutputViewer.SetValue(cSharpCode);
+        }
+
+        async Task ShowRun()
```
Comment (LPeter1997) [inline src/Draco.Editor.Web/App.cs:147]: Accessibility?

File: src/Draco.Editor.Web/App.cs [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+            using (var pe = new PEFile("_", inlineDllStream))
+            {
+                disassembler.WriteAssemblyHeader(pe);
+                text.WriteLine();
+                disassembler.WriteModuleContents(pe);
+            }
+            await this.OutputViewer.SetValue(text.ToString());
+        }
+
+        async Task ShowCSharp()
```
Comment (LPeter1997) [inline src/Draco.Editor.Web/App.cs:139]: Accessibility?

File: src/Draco.Editor.Web/App.cs [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+            default:
+                throw new InvalidOperationException();
+            }
+        }
+        catch (Exception e)
+        {
+            await this.OutputViewer!.SetValue(e.ToString());
+        }
+
+        async Task ShowIL()
```
Comment (LPeter1997) [inline src/Draco.Editor.Web/App.cs:120]: Accessibility?

File: src/Draco.Editor.Web/App.cs [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+            var buffer = inBuffer.ToArray();
+            // Convert base64 to base64URL.
+            var hash = Convert.ToBase64String(buffer)
+                .TrimEnd('=')
+                .Replace('+', '-')
+                .Replace('/', '_');
+            await this.JS.InvokeVoidAsync("setHash", new object[] { hash });
+        }
+
+
```
Comment (LPeter1997) [inline src/Draco.Editor.Web/App.cs:97]: Double stray newlines, one is enough

File: src/Draco.Editor.Web/App.cs [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+    }
+
+    private void CodeChange() => _ = this.UpdateOutput(); // awaiting will block UI thread.
+
+    private async Task UpdateOutput()
+    {
+        var code = await this.DracoEditor!.GetValue();
+        using (var inBuffer = new MemoryStream())
+        using (var outStream = new DeflateStream(inBuffer, CompressionLevel.Optimal, true))
+        using (var writer = new StreamWriter(outStream, Encoding.UTF8, leaveOpen: true))
```
Comment (LPeter1997) [inline src/Draco.Editor.Web/App.cs:82]: Do these need to be block or can they be also inlined?

File: src/Draco.Editor.Web/App.cs [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+        {
+            writer.WriteLine(this.SelectedOutputType);
+            writer.Write(code);
+            writer.Flush();
+            inBuffer.Position = 0;
+            var buffer = inBuffer.ToArray();
+            // Convert base64 to base64URL.
+            var hash = Convert.ToBase64String(buffer)
```
Comment (LPeter1997) [inline src/Draco.Editor.Web/App.cs:90]: Please put the method call to a newline

File: src/Draco.Editor.Web/App.cs [hunk 17]
```diff
@@ -0,0 +1,183 @@
...
+using System.Text;
+
+namespace Draco.Editor.Web;
+
+public partial class App
+{
+    private MonacoEditor? DracoEditor { get; set; }
+    private MonacoEditor? OutputViewer { get; set; }
+    private string? SelectedOutputType { get; set; } = "Run";
+    private static readonly StandaloneEditorConstructionOptions defaultDracoEditorOptions = new()
```
Comment (LPeter1997) [inline src/Draco.Editor.Web/App.cs:18]: Please put an empty line between the props and method.

**Review by LPeter1997 — COMMENTED** [review]
Just a few more annoying notes, sorry.

File: src/Draco.Editor.Web/App.cs (hunk 17 above)
Comment (Kuinox) [reply src/Draco.Editor.Web/App.cs:18]: That's not a method sir.

**Review by LPeter1997 — APPROVED** [review]

**Pull request merged**